sycl/test/basic_tests/half_type.cpp

```diff
@@ -0,0 +1,183 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
+
+#define FLT_EPSILON 9.77e-4
+
+const size_t N = 100;
+
+template <typename T> void assert_close(const T &C, const float ref) {
+  for (size_t i = 0; i < N; i++) {
+    assert(C[i] - ref < FLT_EPSILON);
+  }
+}
+
+void verify_add(queue &q, buffer<half, 1> &a, buffer<half, 1> &b, range<1> &r,
+                const float ref) {
+  buffer<half, 1> c(r);
+
+  q.submit([&](handler &cgh) {
+    auto A = a.get_access<access::mode::read>(cgh);
+    auto B = b.get_access<access::mode::read>(cgh);
+    auto C = c.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class calc_add>(
+        r, [=](id<1> index) { C[index] = A[index] + B[index]; });
+  });
+
+  assert_close(c.get_access<access::mode::read>(), ref);
+}
+
+void verify_min(queue &q, buffer<half, 1> &a, buffer<half, 1> &b, range<1> &r,
+                const float ref) {
+  buffer<half, 1> c(r);
+
+  q.submit([&](handler &cgh) {
+    auto A = a.get_access<access::mode::read>(cgh);
+    auto B = b.get_access<access::mode::read>(cgh);
+    auto C = c.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class calc_min>(
+        r, [=](id<1> index) { C[index] = A[index] - B[index]; });
+  });
+
+  assert_close(c.get_access<access::mode::read>(), ref);
+}
+
+void verify_mul(queue &q, buffer<half, 1> &a, buffer<half, 1> &b, range<1> &r,
+                const float ref) {
+  buffer<half, 1> c(r);
+
+  q.submit([&](handler &cgh) {
+    auto A = a.get_access<access::mode::read>(cgh);
+    auto B = b.get_access<access::mode::read>(cgh);
+    auto C = c.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class calc_mul>(
+        r, [=](id<1> index) { C[index] = A[index] * B[index]; });
+  });
+
+  assert_close(c.get_access<access::mode::read>(), ref);
+}
+
+void verify_div(queue &q, buffer<half, 1> &a, buffer<half, 1> &b, range<1> &r,
+                const float ref) {
+  buffer<half, 1> c(r);
+
+  q.submit([&](handler &cgh) {
+    auto A = a.get_access<access::mode::read>(cgh);
+    auto B = b.get_access<access::mode::read>(cgh);
+    auto C = c.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class calc_div>(
+        r, [=](id<1> index) { C[index] = A[index] / B[index]; });
+  });
+
+  assert_close(c.get_access<access::mode::read>(), ref);
+}
+
+inline bool bitwise_comparison_fp16(const half val, const uint16_t exp) {
+  return *reinterpret_cast<const uint16_t *>(&val) == exp;
+}
+
+inline bool bitwise_comparison_fp32(const half val, const uint32_t exp) {
+  const float fp32 = static_cast<float>(val);
+  return *reinterpret_cast<const uint32_t *>(&fp32) == exp;
```

Comment:
Is there a need to go through memory for this?
I wonder whether `reinterpret_cast<const uint32_t>` could work instead.

Reply:
Compiler doesn't allow me do `reinterpret_cast<const uint32_t>` and I can't find such conversion case on https://en.cppreference.com/w/cpp/language/reinterpret_cast . Only one case without going through memory I could use here : `reinterpret_cast<const uint32_t&>(fp32)` and compiler allows me do it.

Reply:
Yes, you are right you need the reference.